### PR TITLE
PF: Work around flex; make it use gap

### PR DESCRIPTION
**Pull request by garrett**
Local workaround to use flex properly with a gap instead of having a right margin.

Upstream issue: https://github.com/patternfly/patternfly/issues/4523

### CODE REVIEW

**Comment by garrett** [comment]
This fixes the flow wraparound.

**Comment by marusak** [comment]
I think [this](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/pixeldiff.html#TestKeys-testPrivateKeys-ssh-keys-dialog-mobile) might be the real issue for this? That when it wraps, the spacing is correct?
But not sure if [this](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/pixeldiff.html#TestNetworkingBasic-testBasic-network-main-mobile) is intentional, that the checkbox is further from the title. Seen also on [bigger resolution](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/pixeldiff.html#TestNetworkingBasic-testBasic-network-main)

**Review by martinpitt — APPROVED** [review]
Thanks! The [changed pixels](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/log.html) look good to me, I force-pushed the reference updates.

See this [pixel diff comparison](https://github.com/cockpit-project/pixel-test-reference/compare/e294936aa4d21f546e0ae94da7a17578b491c342..97c6f9ddde73bcc3a01b9c62b45f92ee85d6c10e) for the "official" reference change.

**Comment by garrett** [comment]
> I think [this](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/pixeldiff.html#TestKeys-testPrivateKeys-ssh-keys-dialog-mobile) might be the real issue for this? That when it wraps, the spacing is correct? But not sure if [this](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/pixeldiff.html#TestNetworkingBasic-testBasic-network-main-mobile) is intentional, that the checkbox is further from the title. Seen also on [bigger resolution](https://logs.cockpit-project.org/logs/pull-17043-20220224-094403-fd5a366e-fedora-35/pixeldiff.html#TestNetworkingBasic-testBasic-network-main)

The change is intentional. The padding should be handled by gap, not margin. There isn't enough space before. In this case, there's a little more visual space but that's still better than not enough. It's also a comparison of two different resolutions, so the original looks blurry and pixels are a multiplier, so it's not really a great before/after.

Interface action buttons in Networking in mobile is indeed the same problem, but it's with margin-left instead and I think that's a custom Cockpit thing, unrelated to this change, but related to the same problem and solution. (It should be a follow-up for that one.)

Basically this:

image

Should be more like this:

image

It's the result of Safari dragging its feet on accepting gap in anything other than grid (including flex), but I'm pretty sure it's our own one-off compromise instead of PatternFly's.

**Comment by garrett** [comment]
I think you were talking about the switch placement:

image

Which, I think, we have custom spacing to overcome the hacky approach PF has been taking for their Flex element, as it was obviously off in both desktop and mobile.

I think I can craft a couple of quick follow-up patches for these two things, working on the Firewall toggle first.

**Comment by garrett** [comment]
OH! The toggle is because I used unset (which is proper) instead of 0 and we're including CSS more than once or in the wrong order, so it gets re-defined as a space again. :facepalm: 

So it's dueling bugs.

 I can either just set the space to 0, we can detangle the CSS (right approach, but may have side effects and could take a while), or we can have a local redefinition here (hackiest approach).

I suggest we just define the margin to the right as 0. That's the quickest most correct approach we can take without the detangling... which we should definitely do later.

I'm building it locally with 0 instead of unset. Let's see if that fixes it. Thanks for pointing this out, @marusak!

**Comment by garrett** [comment]
I built it with 0 was the margin-right value, but that was still overridden.

Next approach, changing it to:

```scss
    // Remove the margin hack
    > * {
        --pf-l-flex--spacer: 0;
    }
```

This will let margin work if it's defined elsewhere and it should also fix it for the children (but let the value persist properly for the `pf-l-flex` element), thanks to CSS properties (variables) supporting the cascade as well.

I _think_ this should work regardless of any redefinition or ordering issue. Let's see (compiling now).

**Comment by garrett** [comment]
OK, here's the fix. I'll force push it into this PR:

```scss
    // Negate the margin hack used by immediate flex children
    > * {
        --pf-l-flex--spacer-base: 0;
    }
```

It should only affect this one change. The flow of the buttons on the same page should be a follow-up.

**Comment by martinpitt** [comment]
[new changed pixels](https://logs.cockpit-project.org/logs/pull-17043-20220224-120416-9d029b34-fedora-35/log.html) now have less noise, in particular the on/off switch does not change any more. Thanks @garrett! 

Once again I pushed the new reference images, see the [diff view](https://github.com/cockpit-project/pixel-test-reference/compare/e294936aa4d21f546e0ae94da7a17578b491c342..6f9460a4ec2a846d05ac3c5feb6b4129591b3ecb).

**Comment by garrett** [comment]
I'm working on fixing those little bits in generic ways. It's due to PF reassigning variables on children elements instead of the widget itself in this case... which is odd, even though they're doing this in a hacky manner.

I've isolated it to this block of SCSS:

```scss
  // .pf-m-space-items-{size}{-on-breakpoint}
  @each $breakpoint, $breakpoint-value in $pf-l-flex--breakpoint-map {
    $breakpoint-name: if($breakpoint != "base", -on-#{$breakpoint}, "");

    @include pf-apply-breakpoint($breakpoint) {
      @each $spacer, $spacer-value in $pf-l-flex--spacer-map {
        &.pf-m-space-items-#{$spacer}#{$breakpoint-name} {
          > * {
            --pf-l-flex--spacer: var(#{map-get($pf-l-flex--variable-map, $spacer-value)});
          }

          > :last-child {
            --pf-l-flex--spacer: 0;
          }
        }
      }
    }
  }
```

Thankfully, it's not all of it, so I can abstract and simplify it greatly. And this is actually much less complex than it looks.

(Having them switch properly to flex will also _greatly_ simplify their implementation.)

I could probably just handle the `none` case and we can remove the `sm` usage (as it's not needed). Or just handle `none` and `sm`. But I'm looking a bit more at this now.

**Comment by garrett** [comment]
OK, the part I've isolated, when compiled down and further isolated is this:

```css
.pf-l-flex.pf-m-space-items-none > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--none); }
.pf-l-flex.pf-m-space-items-xs > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xs); }
.pf-l-flex.pf-m-space-items-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--sm); }
.pf-l-flex.pf-m-space-items-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--md); }
.pf-l-flex.pf-m-space-items-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--lg); }
.pf-l-flex.pf-m-space-items-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xl); }
.pf-l-flex.pf-m-space-items-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
.pf-l-flex.pf-m-space-items-3xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
.pf-l-flex.pf-m-space-items-4xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--4xl); }

@media (min-width: 577px) {
  .pf-l-flex.pf-m-space-items-none-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--none); }
  .pf-l-flex.pf-m-space-items-xs-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xs); }
  .pf-l-flex.pf-m-space-items-sm-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--sm); }
  .pf-l-flex.pf-m-space-items-md-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--md); }
  .pf-l-flex.pf-m-space-items-lg-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--lg); }
  .pf-l-flex.pf-m-space-items-xl-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xl); }
  .pf-l-flex.pf-m-space-items-2xl-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
  .pf-l-flex.pf-m-space-items-3xl-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--3xl); }
  .pf-l-flex.pf-m-space-items-4xl-on-sm > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--4xl); }
}
@media (min-width: 768px) {
  .pf-l-flex.pf-m-space-items-none-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--none); }
  .pf-l-flex.pf-m-space-items-xs-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xs); }
  .pf-l-flex.pf-m-space-items-sm-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--sm); }
  .pf-l-flex.pf-m-space-items-md-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--md); }
  .pf-l-flex.pf-m-space-items-lg-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--lg); }
  .pf-l-flex.pf-m-space-items-xl-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xl); }
  .pf-l-flex.pf-m-space-items-2xl-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
  .pf-l-flex.pf-m-space-items-3xl-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--3xl); }
  .pf-l-flex.pf-m-space-items-4xl-on-md > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--4xl); }
}
@media (min-width: 992px) {
  .pf-l-flex.pf-m-space-items-none-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--none); }
  .pf-l-flex.pf-m-space-items-xs-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xs); }
  .pf-l-flex.pf-m-space-items-sm-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--sm); }
  .pf-l-flex.pf-m-space-items-md-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--md); }
  .pf-l-flex.pf-m-space-items-lg-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--lg); }
  .pf-l-flex.pf-m-space-items-xl-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xl); }
  .pf-l-flex.pf-m-space-items-2xl-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
  .pf-l-flex.pf-m-space-items-3xl-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--3xl); }
  .pf-l-flex.pf-m-space-items-4xl-on-lg > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--4xl); }
}
@media (min-width: 1200px) {
  .pf-l-flex.pf-m-space-items-none-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--none); }
  .pf-l-flex.pf-m-space-items-xs-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xs); }
  .pf-l-flex.pf-m-space-items-sm-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--sm); }
  .pf-l-flex.pf-m-space-items-md-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--md); }
  .pf-l-flex.pf-m-space-items-lg-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--lg); }
  .pf-l-flex.pf-m-space-items-xl-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xl); }
  .pf-l-flex.pf-m-space-items-2xl-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
  .pf-l-flex.pf-m-space-items-3xl-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--3xl); }
  .pf-l-flex.pf-m-space-items-4xl-on-xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--4xl); }
}
@media (min-width: 1450px) {
  .pf-l-flex.pf-m-space-items-none-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--none); }
  .pf-l-flex.pf-m-space-items-xs-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xs); }
  .pf-l-flex.pf-m-space-items-sm-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--sm); }
  .pf-l-flex.pf-m-space-items-md-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--md); }
  .pf-l-flex.pf-m-space-items-lg-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--lg); }
  .pf-l-flex.pf-m-space-items-xl-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--xl); }
  .pf-l-flex.pf-m-space-items-2xl-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--2xl); }
  .pf-l-flex.pf-m-space-items-3xl-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--3xl); }
  .pf-l-flex.pf-m-space-items-4xl-on-2xl > * { --pf-l-flex--spacer: var(--pf-l-flex--spacer--4xl); }
}
```

There's a ton of duplication for each breakpoint and that's 100% unnecessary. This can be simplified much more. I'm close to a general workaround patch.

**Comment by garrett** [comment]
I've reworked all the spacer modifications like so:

```scss
    // Undo all spacer modification adjustments
    &[class*=pf-m-space-items-] {
        > * {
            --pf-l-flex--spacer: 0;
        }
    }

    // Re-add spacer modification adjustments on the flex layout widget
    // (using class attribute matching for handling breakpoint -on- also)
    @each $size in (none, sm, md, lg, xl, 2xl, 3xl, 4xl) {
        &[class*=pf-m-space-items-#{$size}] {
            --pf-l-flex--spacer-base: var(--pf-l-flex--spacer--#{$size});
        }
    }
```

All we really need currently is `-none` and `-sm`, which wouldn't need the fancy loop and class attribute matching. It would be the simpler form like so:

```scss
    &:where(.pf-m-space-items-none, .pf-m-space-items-md) > * {
        --pf-l-flex--spacer: 0;
    }

    &.pf-m-space-items-none {
        --pf-l-flex--spacer-base: var(--pf-l-flex--spacer--none);
    }

    &.pf-m-space-items-md {
        --pf-l-flex--spacer-base: var(--pf-l-flex--spacer--md);
    }
```

**Comment by martinpitt** [comment]
@garrett: This pixel "failure" [log](https://logs.cockpit-project.org/logs/pull-17043-20220224-152757-ff397763-fedora-35/log.html) is intentional, right? It reverts back to what we have in main.

**Comment by garrett** [comment]
Note: The Health card is broken at the moment:

image

I was trying to fix it. If what I have works, I'll upload it as another commit (as it touches other code), and it'd need review and tweaking.

**Comment by garrett** [comment]
Here's the card with the fixes:

Screenshot 2022-02-24 at 18-18-56 Overview - garrett Bolt

I don't know if all notifications align or not though, and if this is different from main or not. I haven't had time to check. The spacer size is sm; it might need to be changed otherwise if it's not correct.

The biggest point is that I moved the spacer from a `spaceSm` in the first `FlexItem` (for the icon) to a `spaceItemsSm` in the parent `Flex` for the two items in 4007700.

**Comment by garrett** [comment]
> @garrett: This pixel "failure" [log](https://logs.cockpit-project.org/logs/pull-17043-20220224-152757-ff397763-fedora-35/log.html) is intentional, right? It reverts back to what we have in main.

No, everything should be fixed now. I specifically checked that one too.

It still looks fine here (as much as it can — I don't really like this UI, but that's a different story):

image

**Comment by martinpitt** [comment]
Thanks @garrett! I reverted the previous pixel ref update for the ssh key dialog, as this now does not change any more. (Force push alert)

**Review by martinpitt — APPROVED** [review]
